### PR TITLE
Session billing integration

### PR DESCRIPTION
**Pull request by LukasPaczos**
### Description

This PR integrates session billing for the Navigation SDK. More billing documentation and rollout plan information is to follow.

I'm cutting this PR early to start the integration review process while we still wait for other dependencies. @mapbox/navigation-android I'd appreciate a round of reviews if anyone has the bandwidth.

The biggest change not directly related to the billing system is the fact that we're now enforcing a single instance of `MapboxNavigation` in an application process, where the lifecycle is defined from object creation to `MapboxNavigation#onDestroy` call. This was always how we documented the usage of the Nav SDK, now this is additionally verified in the runtime. This is critical for being able to bill correctly since the `BillingService` is a singleton.

### Changelog

```
<changelog>:exclamation: Introduced new pricing options. Integrating this or a later version of the Navigation SDK will impact the way you are billed. Read more in our [Pricing Documentation](https://docs.mapbox.com/android/beta/navigation/guides/pricing/).</changelog>
<changelog>:warning: `MapboxNavigation` now enforces having only one instance alive in a process, a new instance cannot be created if there's another one that did not have `#onDestroy` called. Use `MapboxNavigationProvider` for assistance with instance management.</changelog>
```

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+        return true
+    }
+
+    private fun handlerError(error: BillingServiceError) {
+        when (error.code) {
+            BillingServiceErrorCode.INVALID_SKU_ID,
+            BillingServiceErrorCode.RESUME_FAILED,
+            null -> {
+                throw IllegalArgumentException(error.toString())
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt:189]: Maybe better? 🤔 

```suggestion
                throw IllegalStateException(error.toString())
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                NavigationSession.State.FREE_DRIVE -> {
+                    BillingServiceWrapper.beginBillingSession(
+                        SKUIdentifier.NAV2_SES_TRIP,
+                        validity = TimeUnit.HOURS.toMillis(1) // validity of 1hr
+                    )
+                }
+                NavigationSession.State.ACTIVE_GUIDANCE -> {
+                    BillingServiceWrapper.beginBillingSession(
+                        SKUIdentifier.NAV2_SES_TRIP,
+                        validity = 0 // default validity, 12hrs
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt:58]: Wouldn't be clear to pass `12`hrs (and remove the comment)?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            }
+
+            when (navigationSessionState) {
+                NavigationSession.State.IDLE -> {
+                    // do nothing
+                }
+                NavigationSession.State.FREE_DRIVE -> {
+                    BillingServiceWrapper.beginBillingSession(
+                        SKUIdentifier.NAV2_SES_TRIP,
+                        validity = TimeUnit.HOURS.toMillis(1) // validity of 1hr
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt:52]:
```suggestion
                        validity = TimeUnit.HOURS.toMillis(1)
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+            )
+            val newWaypoints = getWaypointsOnRoute(directionsRoute)
+
+            if (!waypointsWithinRange(currentRemainingWaypoints, newWaypoints)) {
+                // stop any running sessions
+                BillingServiceWrapper.stopBillingSession(SKUIdentifier.NAV2_SES_TRIP)
+
+                BillingServiceWrapper.beginBillingSession(
+                    SKUIdentifier.NAV2_SES_TRIP,
+                    validity = 0 // default validity, 12hrs
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt:106]: Same as above

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 5]
```diff
@@ -414,6 +436,18 @@ class MapboxNavigation(
         navigator.resetRideSession()
     }
 
+    fun pauseTripSession() {
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:439]: Does this need to be `public`?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 6]
```diff
@@ -414,6 +436,18 @@ class MapboxNavigation(
...
+        runIfNotDestroyed {
+            billingController.pauseSession()
+        }
+    }
+
+    fun resumeTripSession() {
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:445]: Same as above

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 7]
```diff
@@ -273,7 +274,20 @@ class MapboxNavigation(
 
     private var reachabilityObserverId: Long? = null
 
+    private var isDestroyed = false
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:277]: Do we really need `isDestroyed`? Apparently `hasInstance` has the same lifecycle 🤔

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/BillingControllerTest.kt [hunk 8]
```diff
@@ -0,0 +1,575 @@
...
+                any(),
+                0
+            )
+        }
+    }
+
+    @Test
+    fun `when pause session, native APIs called`() {
+        billingController.pauseSession()
+        verify(exactly = 1) {
```
Comment (Guardiola31337) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/BillingControllerTest.kt:150]: NIT

```suggestion

        verify(exactly = 1) {
```

**Review by Guardiola31337 — COMMENTED** [review]
Left some initial comments.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt (hunk 2 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt:58]: We don't want to limit active guidance sessions at all. 12 hours is just a server-side limit that might be lifted at some point. I can remove the `12hrs` note and just indicate that this is max possible duration (whatever that is at a certain point in time).

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt (hunk 5 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:439]: Yes, this has to be invoked by developers when preferred. Docs to be added.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt (hunk 2 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt:58]: Sounds good 👍 

I'm assuming `0` is the "max" then, right? That's why it's confusing 😅

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/BillingController.kt:58]: That's right, it is indeed a tiny bit confusing.

**Comment by LukasPaczos** [comment]
This is ready for a final round of reviews.

**Review by Guardiola31337 — APPROVED** [review]
:shipit:

**Pull request merged**